### PR TITLE
Multi Stage Training

### PR DESCRIPTION
**Pull request by jotix16**
This PR introduces a Multi-Stage training logic via a class called MultiStager.
MultiStager is a wrapper around Pretrain. It enables multi-stage training which is as far as I know only useful for transducer like models. The idea is to train interchangeable using FullSum loss and CrossEntropy loss. For the CE loss the alignment are created by the FS training.

Since we already had most of the parts from the RNA training config, I just put it together.  Right now I am mostly using functions for the RNA topology but we can extend that with RNN-T or CTC topology.
The usage in a config file could look like 
``` python
get_network = MultiStager([
          Pretrain( make_net_FS, {"enc_lstm_dim": (512, 1024), "enc_num_layers": (3, 6)}, num_epochs=50),
          Pretrain( make_net_CE, {"enc_lstm_dim": (512, 1024), "enc_num_layers": (3, 6)}, num_epochs=20),
          Pretrain( make_net_FS, {"enc_lstm_dim": (1024, 1024), "enc_num_layers": (6, 6)}, num_epochs=50),
          Pretrain( make_net_CE, {"enc_lstm_dim": (512, 1024), "enc_num_layers": (3, 6)}, num_epochs=20),
        ]).get_network
```

Here are some stuff I still need to clarify

TODO:
  - [ ] Make sure that the alignments correspond to the dataset used(sequence_ordering, ..)
  - [ ] Are there so many configuration differences between FS and CE network? If not we could use
        a function like self.make_ce() which adds the required differences just like self.make_align().
  - [ ] How to get targetb_num_labels, how to get epoch_split? Dataset?
  - [ ] Reset option, for example we can train with CE for a while and than switch to FS with same weights

### CODE REVIEW

**Comment by albertz** [comment]
Hey!

(Sorry for the delay, we had Interspeech submission deadline.)

To simplify the review, can you split up this PR? E.g. things like `rna_code/rna_tf_impl.py` should better go into a separate PR. (And as they are just copy & pasted from earlier code, I think we mostly just can overtake them.) But then we can better focus in this PR about the real important things, or those things which maybe need some discussion.

**Comment by albertz** [comment]
> MultiStager is a wrapper around Pretrain

Why around Pretrain? It rather looks like this would be more clean if it is a wrapper or combiner for some generic function / API which defines a `get_network` function `epoch -> network`. Right?

**Comment by albertz** [comment]
> `MultiStager.make_align`

I think it would be more clean if the `make_align` is more separated. It could be a `MakeAlignment` class (which also maybe implements a `GetNetwork` API / base interface), and which also provides the HDF files of the alignment.

**Comment by albertz** [comment]
I assume you have now an idea of how the original full pipeline setup works (see also #311 for initial discussions on that).

However, we do not necessarily need to stick to that. We can extend it however we want. Do not assume that we have the optimal setup, and try to wrap that up somehow. Rather, think more generally, how a clean setup would/should look like, design the setup, and then maybe we need to make RETURNN more flexible to allow for that.

Or maybe you think it's already optimal the way it is, or you don't have a better idea.

As also discussed in #311 (and maybe elsewhere), in general, we need some good/simple/robust/clean way to store auxiliary/extra information. In this case it is an alignment (per seq from the dataset). Maybe there are other things in the future (e.g. I'm thinking about RL, where you might want to store rollouts). And this information might get updated (e.g. in some later epochs, we would want to update the alignment).

We maybe want to apply the same `cleanup_old_models` logic on these extra information. (This is not really possible right now, because RETURNN doesn't treat generated files with `HDFDumpLayer` this way.) But maybe also not always? Or not exactly the same logic, but similar (keep last N or so)? Maybe also not so important, as this should be so huge usually? (But this assumption might be wrong in other cases.)

Do we actually correctly use cache manager for these generated files?

Then, after generated, we need a good generic way to access the data (the most recent file, or most recent N of full epoch, or maybe more?). In the full pipeline config, this is all kind of manual/explicitly handled, which might be prone to bugs, is complicated and not so straight-forward. And e.g. for multi-GPU training, I'm not even sure if it does the correct thing.

So, instead of wrapping up just my earlier setup for returnn-experiments `common`, maybe let's take a step back and think a bit about if we can maybe improve it somehow, so it becomes simpler, more robust.

**Comment by jotix16** [comment]
> To simplify the review, can you split up this PR? E.g. things like rna_code/rna_tf_impl.py should better go into a separate PR
Yes will just copy them from the old configs and make a new PR.

>Why around Pretrain? It rather looks like this would be more clean if it is a wrapper or combiner for some generic function / API which defines a get_network function epoch -> network. Right?
Exactly, it could be an arbitrary function that gets an epoch and returns a network. The general case is that every stage starts with Pretraining. For that the Pretrainer needs to know the start epoch of the stage. We could initilaize it in the lazy_init of the get_network function of the Pretrainerr:
```python
def _lazy_init(self, epoch):
    if self._is_initialized:
      return
    self._is_initialized = True

    config = get_global_config()
    self._start_epoch = epoch
    self._lr_std = config.typed_dict["learning_rate"]
    self._lr_warmup_initial = self._lr_std / 10.
```

**Comment by albertz** [comment]
> > Why around Pretrain? It rather looks like this would be more clean if it is a wrapper or combiner for some generic function / API which defines a get_network function epoch -> network. Right?
> 
> Exactly, it could be an arbitrary function that gets an epoch and returns a network. The general case is that every stage starts with Pretraining.

Why Pretraining? As you say, it could be an arbitrary function. So why does it need to be pretraining? Our terminology of pretraining implies that this is only done in the beginning, with is a contradiction here, so pretraining doesn't really make sense.

So, as you say, let it be an arbitrary function. Or arbitrary interface (class). The pretraining can also derive from that interface as a special case. But this multistager should not need to know anything about pretraining.

> For that the Pretrainer needs to know the **start epoch** of the stage.

Why? Pretraining is meant to start in the beginning. If it does not start in the beginning, this is not pretraining.

> We could initilaize it in the lazy_init of the get_network function of the Pretrainer.

Ignore the current code (always). Think about what we want, how it should be, and then we can think about what needs to be changed.

**Comment by jotix16** [comment]
>So why does it need to be pretraining? Our terminology of pretraining implies that this is only done in the beginning, with is a contradiction her 

Depending on the situation every new stage can be seen as a new training, i.e. in the beginning of this new training we can use pretraining.

> So, as you say, let it be an arbitrary function. Or arbitrary interface (class). The pretraining can also derive from that interface as a special case. But this multistager should not need to know anything about pretraining.

yes. Pretraining is rather an Add-On on this arbitrary function. So there is no need to differentiate a simple function that just returns a network from a more advanced function that returns a network considering some pretraining logic, as long as it takes an epoch as input and returns the network dictionary.

>Why? Pretraining is meant to start in the beginning. If it does not start in the beginning, this is not pretraining.

Just wanted to express that defining `beginning` as epoch=1 or 0 is not the right way. For example, epoch=1 or 0 can be used to do some fancy forward pass and initialise the network and after that pretraining can start. 

>As also discussed in #311 (and maybe elsewhere), in general, we need some good/simple/robust/clean way to store auxiliary/extra information 

I can grasp now what you meant in [this comment](https://github.com/rwth-i6/returnn-experiments/pull/60#issuecomment-814877760). While solving the problem why not solve it in general for different auxiliary information we want to save during/after a training stage. So rather the logic in what form this information is to be saved. Right?
I actually took it for granted that `HDFDumplayer` is the way and I only have to build the logic around it.

>Ignore the current code (always). Think about what we want, how it should be, and then we can think about what needs to be changed.

**what we want**  
At least for CTC/Transducer like problems we marginalize over alignments which we ourself predict. FullSum is the right way but maybe as explored in [zeyer 2020](https://www-i6.informatik.rwth-aachen.de/publications/download/1145/ZeyerAlbertMerboldtAndr%E9Schl%FCterRalfNeyHermann--ANewTrainingPipelineforanImprovedNeuralTransducer--2020.pdf), using forced alignments with CE training can improve training stability and open up different possibilities to include training tricks, such as chunking, label smoothing, etc. 

We construct/define the network/s as dictionaries that do the default FullSum training and upgrade the training with MultiStager that automises:
- The creation of alignments for any ctc/transducer like model, so that we don't have to think about this while constructing the network.
- The exploition of these alignments for CE training of a given model
- Maybe some logic to allow iterating between FS and CE for the same model without starting training from the beginning. That of course should allow the usage of a Pretrain like routine that expands the network.

TLDR
MultiStage training can just be used on top of the normal training without having to think about it during the network's construction.

**Comment by albertz** [comment]
> > So, as you say, let it be an arbitrary function. Or arbitrary interface (class). The pretraining can also derive from that interface as a special case. But this multistager should not need to know anything about pretraining.
> 
> yes. Pretraining is rather an Add-On on this arbitrary function. So there is no need to differentiate a simple function that just returns a network from a more advanced function that returns a network considering some pretraining logic, as long as it takes an epoch as input and returns the network dictionary.

But it should not be called "pretraining". Pretraining is a special case. It should be called `MakeNetwork` or `GetNetwork` or so.

> > Why? Pretraining is meant to start in the beginning. If it does not start in the beginning, this is not pretraining.
> 
> Just wanted to express that defining `beginning` as epoch=1 or 0 is not the right way. For example, epoch=1 or 0 can be used to do some fancy forward pass and initialise the network and after that pretraining can start.

But it could be relative epoch 0 for this stage. Why should the stage need to know what absolute epoch it is? That logic can be left to the MultiStager, such that the MultiStager would call `get_network(stage_start_epoch + stage_rel_epoch)` or so.

> > As also discussed in #311 (and maybe elsewhere), in general, we need some good/simple/robust/clean way to store auxiliary/extra information
> 
> I can grasp now what you meant in [this comment](https://github.com/rwth-i6/returnn-experiments/pull/60#issuecomment-814877760). While solving the problem why not solve it in general for different auxiliary information we want to save during/after a training stage. So rather the logic in what form this information is to be saved. Right?
> I actually took it for granted that `HDFDumplayer` is the way and I only have to build the logic around it.

(Oh, I realize this is here in returnn-experiments. I meant [returnn #311](https://github.com/rwth-i6/returnn/issues/311).)

Yes exactly, it should be generic. Actually `HDFDumpLayer` is already generic in that it stores whatever you give it. But this is still specific to cases where you have aux info per sequence. Maybe there are other cases where you accumulate statistics over multiple seqs, part of the corpora. Or so. But maybe we can ignore this for now, and assume we want aux info per seq. And maybe also only the most recent such info, and earlier aux info can be deleted.

`HDFDumpLayer` doesn't really have much (any) special logic, though. It just dumps to a HDF file. There is no clever logic to handle that for multiple epochs. I.e. currently you need to make sure that you dump to different HDF files each epoch. And this is also completely separate from the dataset loading. You decide what you do with those HDF files, and how you use them. They will not be loaded automatically in the next epoch or so.

This is all doable (as you see in the transducer full training pipeline config), but somewhat complicated, and easy to get wrong. Maybe we can simplify it somehow? Maybe on RETURNN side?

I don't have a solution, nor idea really. I think this needs some more thought.

> **what we want**
> At least for CTC/Transducer like problems we marginalize over alignments which we ourself predict. FullSum is the right way but maybe as explored in [zeyer 2020](https://www-i6.informatik.rwth-aachen.de/publications/download/1145/ZeyerAlbertMerboldtAndr%E9Schl%FCterRalfNeyHermann--ANewTrainingPipelineforanImprovedNeuralTransducer--2020.pdf), using forced alignments with CE training can improve training stability and open up different possibilities to include training tricks, such as chunking, label smoothing, etc.

Yes right, but this is now specific for this example.

> We construct/define the network/s as dictionaries that do the default FullSum training and upgrade the training with MultiStager that automises:
> 
> * The creation of alignments for any ctc/transducer like model, so that we don't have to think about this while constructing the network.

I don't exactly understand. Constructing what network? Constructing the network for CE? Constructing the network for FS? Constructing the network for creating the alignments?

> * The exploition of these alignments for CE training of a given model

How does the MultiStager automizes this?

(Note that it is also not really "exploitation". This is just how framewise CE is done normally. This is the standard way.)

> * Maybe some logic to allow iterating between FS and CE for the same model without starting training from the beginning. That of course should allow the usage of a Pretrain like routine that expands the network.

This is not a "maybe". This should be one core functionality. This is one of the first things we should study now.

(It looks like FS -> Align -> CE is too simple, and does not work good enough e.g. on Librispeech. Or maybe FS was too short. But I think FS -> Loop( Align -> CE ) might be better. Or maybe FS -> Loop( Align -> CE -> approx-FS).)

(Alignment creation / dumping can also be done on-the-fly while training with CE/FS. That might be faster, because you already have all in memory, and also already calculated the encoder. Although encoder has used dropout, and maybe alignment is better without dropout? Unclear...)

Note that expanding the network (or changing whatever in it) is always possible. This doesn't need to be pretraining.

> TLDR
> MultiStage training can just be used on top of the normal training without having to think about it during the network's construction.

Yes sure, that's a basic rule to follow for all abstractions.
